aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -68,37 +54,139 @@ template <> struct DistAccumType<double> { using type = double; };
 template <typename T>
 using dist_acctype = typename DistAccumType<T>::type;
 
-// Constants for uniform distribution
-// doubles have 52 bits of mantissa (fractional part)
-constexpr uint64_t DOUBLE_MASK = (1ULL << 53) - 1;
-constexpr double DOUBLE_DIVISOR = 1.0 / (1ULL << 53);
+/**
+ * Samples a discrete uniform distribution in the range [base, base+range) of type T
+ * This is a transformation function for
+ * https://pytorch.org/docs/stable/tensors.html?highlight=random#torch.Tensor.random_
+ * when both `from` and `to` are specified.
+ */
+template <typename T>
+struct uniform_int_from_to_distribution {
+
+  C10_HOST_DEVICE inline uniform_int_from_to_distribution(uint64_t range_, int64_t base_) {
+    range = range_;
+    base = base_;
+  }
+
+  template <typename V,
+            typename std::enable_if<std::is_integral<V>::value, int>::type = 0>
+  C10_HOST_DEVICE inline T operator()(V val) {
```

Comment:
So, if I understand correctly, these classes are modeled after the `std::uniform_int_distribution` classes in the standard library. This seems reasonable. But my understanding of how those classes worked was that the `operator()` overloads always accept the RNG source as an argument. However, here you have overloads that accept some arbitrary integral V argument. That's pretty weird. What is that supposed to mean? Is a user supposed to ever make use of this?

(I'm guessing the interface looks this way because on CUDA you can't actually pass in the generator? But in that case I would expect some more safety to make sure you pass in enough entropy, i.e., a big enough random number, to actually properly compute the intended random number. And I would definitely expect this to be documented one way or another.)

Reply:
I moved what I called "transformation functions" to TransformationHelper.h